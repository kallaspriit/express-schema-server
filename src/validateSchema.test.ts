import { ICustomValidator, JSONSchema4, validateJsonSchema } from "./index";

// returns valid password for foobar, false otherwise
const validPassword = "foobar";
const validEmail = "foo@bar.com";
const validatePassword: ICustomValidator = {
  name: "valid-password",
  validate: async (password: string) => password === validPassword,
};
const schema: JSONSchema4 = {
  title: "Login",
  description: "Authenticate with username and password",
  type: "object",
  properties: {
    email: {
      type: "string",
      title: "Email",
      description: "Email address",
      minLength: 3,
      maxLength: 256,
      format: "email",
    },
    password: {
      type: "string",
      title: "Password",
      description: "Account password",
      minLength: 6,
      maxLength: 256,
      format: "valid-password",
    },
  },
  required: ["email", "password"],
};
const validators = [validatePassword];

// tslint:disable:no-magic-numbers
describe("express-schema-server", () => {
  it("validates json schema using both built-in and custom validators", async () => {
    const data = {
      email: validEmail,
      password: validPassword,
    };
    const result = await validateJsonSchema(data, schema, validators);

    expect(result.isValid).toEqual(true);
  });

  it("reports missing fields", async () => {
    const data = {
      email: validEmail,
    };
    const result = await validateJsonSchema(data, schema, validators);

    expect(result.isValid).toEqual(false);
    expect(result.errors.length).toEqual(1);
    expect(result.errors).toMatchSnapshot();
  });

  it("reports all issues", async () => {
    const data = {
      email: "",
      password: "",
    };
    const result = await validateJsonSchema(data, schema, validators);

    expect(result.isValid).toEqual(false);
    expect(result.errors.length).toEqual(3);
    expect(result.errors).toMatchSnapshot();
  });

  it("reports invalid custom validator", async () => {
    const data = {
      email: validEmail,
      password: "xxxxxx",
    };
    const result = await validateJsonSchema(data, schema, validators);

    expect(result.isValid).toEqual(false);
    expect(result.errors.length).toEqual(1);
    expect(result.errors).toMatchSnapshot();
  });

  it("reports built in validator along with custom custom validator", async () => {
    const data = {
      email: "yyy",
      password: "xxx",
    };
    const result = await validateJsonSchema(data, schema, validators);

    expect(result.isValid).toEqual(false);
    expect(result.errors.length).toEqual(2);
    expect(result.errors).toMatchSnapshot();
  });
});
