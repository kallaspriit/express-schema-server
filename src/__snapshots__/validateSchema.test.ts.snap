// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`express-schema-server reports all issues 1`] = `
Array [
  Object {
    "code": "MIN_LENGTH",
    "description": "Account password",
    "message": "String is too short (0 chars), minimum 6",
    "params": Array [
      0,
      6,
    ],
    "path": "#/password",
    "schemaId": undefined,
    "title": "Password",
    Symbol(z-schema/schema): Object {
      "__$validated": true,
      "description": "Account password",
      "format": "valid-password",
      "maxLength": 256,
      "minLength": 6,
      "title": "Password",
      "type": "string",
    },
    Symbol(z-schema/json): Object {
      "email": "",
      "password": "",
    },
  },
  Object {
    "code": "MIN_LENGTH",
    "description": "Email address",
    "message": "String is too short (0 chars), minimum 3",
    "params": Array [
      0,
      3,
    ],
    "path": "#/email",
    "schemaId": undefined,
    "title": "Email",
    Symbol(z-schema/schema): Object {
      "__$validated": true,
      "description": "Email address",
      "format": "email",
      "maxLength": 256,
      "minLength": 3,
      "title": "Email",
      "type": "string",
    },
    Symbol(z-schema/json): Object {
      "email": "",
      "password": "",
    },
  },
  Object {
    "code": "INVALID_FORMAT",
    "description": "Email address",
    "message": "Object didn't pass validation for format email: ",
    "params": Array [
      "email",
      "",
    ],
    "path": "#/email",
    "schemaId": undefined,
    "title": "Email",
    Symbol(z-schema/schema): Object {
      "__$validated": true,
      "description": "Email address",
      "format": "email",
      "maxLength": 256,
      "minLength": 3,
      "title": "Email",
      "type": "string",
    },
    Symbol(z-schema/json): Object {
      "email": "",
      "password": "",
    },
  },
  Object {
    "code": "INVALID_FORMAT",
    "description": "Account password",
    "message": "Object didn't pass validation for format valid-password: ",
    "params": Array [
      "valid-password",
      "",
    ],
    "path": "#/",
    "schemaId": null,
    "title": "Password",
    Symbol(z-schema/schema): Object {
      "__$validated": true,
      "description": "Account password",
      "format": "valid-password",
      "maxLength": 256,
      "minLength": 6,
      "title": "Password",
      "type": "string",
    },
    Symbol(z-schema/json): Object {
      "email": "",
      "password": "",
    },
  },
]
`;

exports[`express-schema-server reports built in validator along with custom custom validator 1`] = `
Array [
  Object {
    "code": "MIN_LENGTH",
    "description": "Account password",
    "message": "String is too short (3 chars), minimum 6",
    "params": Array [
      3,
      6,
    ],
    "path": "#/password",
    "schemaId": undefined,
    "title": "Password",
    Symbol(z-schema/schema): Object {
      "__$validated": true,
      "description": "Account password",
      "format": "valid-password",
      "maxLength": 256,
      "minLength": 6,
      "title": "Password",
      "type": "string",
    },
    Symbol(z-schema/json): Object {
      "email": "yyy",
      "password": "xxx",
    },
  },
  Object {
    "code": "INVALID_FORMAT",
    "description": "Email address",
    "message": "Object didn't pass validation for format email: yyy",
    "params": Array [
      "email",
      "yyy",
    ],
    "path": "#/email",
    "schemaId": undefined,
    "title": "Email",
    Symbol(z-schema/schema): Object {
      "__$validated": true,
      "description": "Email address",
      "format": "email",
      "maxLength": 256,
      "minLength": 3,
      "title": "Email",
      "type": "string",
    },
    Symbol(z-schema/json): Object {
      "email": "yyy",
      "password": "xxx",
    },
  },
  Object {
    "code": "INVALID_FORMAT",
    "description": "Account password",
    "message": "Object didn't pass validation for format valid-password: xxx",
    "params": Array [
      "valid-password",
      "xxx",
    ],
    "path": "#/",
    "schemaId": null,
    "title": "Password",
    Symbol(z-schema/schema): Object {
      "__$validated": true,
      "description": "Account password",
      "format": "valid-password",
      "maxLength": 256,
      "minLength": 6,
      "title": "Password",
      "type": "string",
    },
    Symbol(z-schema/json): Object {
      "email": "yyy",
      "password": "xxx",
    },
  },
]
`;

exports[`express-schema-server reports invalid custom validator 1`] = `
Array [
  Object {
    "code": "INVALID_FORMAT",
    "description": "Account password",
    "message": "Object didn't pass validation for format valid-password: xxxxxx",
    "params": Array [
      "valid-password",
      "xxxxxx",
    ],
    "path": "#/",
    "schemaId": null,
    "title": "Password",
    Symbol(z-schema/schema): Object {
      "__$validated": true,
      "description": "Account password",
      "format": "valid-password",
      "maxLength": 256,
      "minLength": 6,
      "title": "Password",
      "type": "string",
    },
    Symbol(z-schema/json): Object {
      "email": "foo@bar.com",
      "password": "xxxxxx",
    },
  },
]
`;

exports[`express-schema-server reports missing fields 1`] = `
Array [
  Object {
    "code": "OBJECT_MISSING_REQUIRED_PROPERTY",
    "description": "Authenticate with username and password",
    "message": "Missing required property: password",
    "params": Array [
      "password",
    ],
    "path": "#/",
    "schemaId": undefined,
    "title": "Login",
    Symbol(z-schema/schema): Object {
      "__$compiled": true,
      "__$validated": true,
      "description": "Authenticate with username and password",
      "properties": Object {
        "email": Object {
          "__$validated": true,
          "description": "Email address",
          "format": "email",
          "maxLength": 256,
          "minLength": 3,
          "title": "Email",
          "type": "string",
        },
        "password": Object {
          "__$validated": true,
          "description": "Account password",
          "format": "valid-password",
          "maxLength": 256,
          "minLength": 6,
          "title": "Password",
          "type": "string",
        },
      },
      "required": Array [
        "email",
        "password",
      ],
      "title": "Login",
      "type": "object",
    },
    Symbol(z-schema/json): Object {
      "email": "foo@bar.com",
    },
  },
]
`;
